mps2_an521_nonsecure: set up running samples and tests in CI

ioannisg opened the pull request:
Allow samples and tests tagged with `-arm` and `-kernel` tags to build and run for Cortex-M33 Non-Secure. The QEMU target is `mps2_an521_nonsecure`.
- this will prevent developers from introducing new or modifying existing test-cases that - by design - cannot run on non-secure domains
- the focus is on Cortex-M specific ARCH tests, as well as userspace tests.
- Non-secure builds need to be combined with TF-M as the secure component; this is set up by default on mps2_an521_nonsecure

ioannisg committed:
boards: mps2_an521: build non-secure Zephyr images together with TF-M

By default, if we build for a Non-Secure version of the board,
force building with TF-M as the Secure Execution Environment.
An exception is when we build openamp samples, where the non-
secure image is basically used for the remote core.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: mps2_an521: add arm, kernel and userspace tags for nonsecure

Build and execute tests tagged with -arm, -kernel,
and -userspace, for mps2_an521_nonsecure.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented on boards/arm/mps2_an521/board.cmake:
```diff
@@ -11,3 +11,10 @@ set(QEMU_FLAGS_${ARCH}
+  # Override the binary used by qemu, to use the combined
+  # TF-M (Secure) & Zephyr (Non Secure) image (when running
+  # in-tree tests).
+  set(QEMU_KERNEL_OPTION "-device;loader,file=${CMAKE_BINARY_DIR}/tfm_merged.hex")
```
@tejlmand FYI

galak replied:
Any implications to this getting set in `samples/tfm_integration/tfm_ipc/CMakeLists.txt` and `samples/tfm_integration/psa_level_1/CMakeLists.txt`?

oyvindronningstad replied:
I guess this should be removed from the other places now.

ioannisg replied:
> I guess this should be removed from the other places now.

Well, sure. I kept it there, for the samples to be self-sustained, i.e. have the qemu kernel option set, independent of the board setting. But yeap, it can be removed.

ioannisg replied:
@galak that's the point, actually. I would like all the samples and tests to run with tfm, so this setting needs to be generic...

tejlmand replied:
@ioannisg not directly related to this PR, but I notice we have both:
https://github.com/zephyrproject-rtos/zephyr/blob/8697935882a8daffe63904464ace70a3f4d77200/modules/trusted-firmware-m/CMakeLists.txt#L322-L331

and:
https://github.com/zephyrproject-rtos/zephyr/blob/8697935882a8daffe63904464ace70a3f4d77200/CMakeLists.txt#L1389-L1402

Could you create a task on aligning those, including making the `merged.hex` / `tfm_merged.hex` name configurable like `KERNEL_BIN_NAME` is:
https://github.com/zephyrproject-rtos/zephyr/blob/8697935882a8daffe63904464ace70a3f4d77200/Kconfig.zephyr#L274-L278

galak commented on boards/arm/mps2_an521/Kconfig.defconfig:
```diff
@@ -15,6 +15,13 @@ config BOARD
+
+config BUILD_WITH_TFM
+	default y if TRUSTED_EXECUTION_NONSECURE
+
```
Should we be do this for all "nonsecure" board configs?

ioannisg replied:
I do not have a strong opinion.
I did this for the qemu target, as TFM is the "official" SPE solution for Zephyr. But, I would not "force" the other vendors (STM, NXP) to use TFM by default when building their non-secure applications.

ioannisg commented:
There's only ztest/error_hook failing (which i am sure I had fixed for nrf5340_cpuapp_ns, but perhaps there's another failure in the qemu that needs to be fixed).

ioannisg commented:
> There's only ztest/error_hook failing (which i am sure I had fixed for nrf5340_cpuapp_ns, but perhaps there's another failure in the qemu that needs to be fixed).

I think the problem is that QEMU does not properly emulate access to private Vendor_SYS. I'll investigate, or disable this particular  test case for Non-Secure QEMU.

galak commented:
> I think the problem is that QEMU does not properly emulate access to private Vendor_SYS. I'll investigate, or disable this particular test case for Non-Secure QEMU.

What is Vendor_SYS?

ioannisg commented:
> > I think the problem is that QEMU does not properly emulate access to private Vendor_SYS. I'll investigate, or disable this particular test case for Non-Secure QEMU.
> 
> What is Vendor_SYS?

Vendor System memory region. This area is usually Exempted from security checking, so accessing memory areas here does not trigger SecureFaults. In nRF accessing 0xFFFFFFFF from non-secure works fine, but apparently this is not the case for MPS2 AN521. @galak the test can be easily made to work for AN521 by selecting some other kernel memory to implement access fault (instead of 0xFFFFFFFF). I'll add a patch.

ioannisg committed:
tests: error_hook: fix test case for non-secure cortex-m platforms

When running the test on Cortex-M platforms, use access to
_current to trigger a memory access fault, as this address
is guaranteed to be in kernel ram and will not trigger
SecureFault exceptions.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
Fixed the error hook test, making the fault trigger test case more generic for cortex m. I am only wondering now, whether I should also enable the `-kernel` tag, for the non secure AN521 qemu, besides `-arm` and `-userspace`. @nashif @galak any thoughts?

ioannisg commented:
Rebased once more and enabled the -kernel tag for the mps2_an521_nonsecure. That should give good enough CI coverage for Non-Secure cortex-m , imho.

ioannisg committed:
doc: releases: add note about non-secure cortex-m QEMU runs in CI

Add a note in the 2.6 release notes about configuring QEMU
to run in CI on mps2_an521_nonsecure with TF-M as the secure
firmware component.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
Release notes entry added.

carlescufi reviewed (approved):
LGTM
